devicetree: extern dt device structs

galak opened the pull request:
Add extern of possible device struct based on DT so we don't have to put
a DEVICE_DT_DECLARE everywhere.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
devicetree: generate extern's for devicetree struct devices

Generate a header (device_extern.h) that handles extern of possible
device structs that would come from devicetree.  This removes the need
for DEVICE_DT_DECLARE and DEVICE_DT_INST_DECLARE which we can remove.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic reviewed (changes requested):
As discussed on Slack, this doesn't belong in devicetree.h. The device layer is higher up.

galak commented:
Moved to creating a separate file `device_extern.h`

pabigot commented on scripts/dts/gen_defines.py:
```diff
@@ -110,6 +110,28 @@ def main():
+        print("", file=dev_header_file)
+
+        for node in sorted(edt.nodes, key=lambda node: node.dep_ordinal):
+            print(f"extern const struct device __device_DT_{node.z_path_id};", file=dev_header_file)
```
I'd rather not have this hard-code the representation of the external symbol.  Can we move the include of the generated file to the end of `device.h` and have this just use `DT_DEVICE_DECLARE();`?

galak replied:
I like that, will rework this to use DT_DEVICE_DECLARE().

pabigot reviewed (changes requested):
Don't want this hard-coding what a maintained macro generates.

mbolivar-nordic commented on include/device.h:
```diff
@@ -21,6 +21,7 @@
 
 #include <init.h>
 #include <sys/device_mmio.h>
+#include <device_extern.h>
```
Maybe comment that this is a generated file?

galak replied:
Added comment.

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -110,6 +110,28 @@ def main():
+        print("#ifndef DEVICE_EXTERN_GEN_H", file=dev_header_file)
+        print("#define DEVICE_EXTERN_GEN_H", file=dev_header_file)
+        print("", file=dev_header_file)
+        print("#ifndef _ASMLANGUAGE", file=dev_header_file)
```
Not needed since it's in device.h now.

galak replied:
done.

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -110,6 +110,28 @@ def main():
         write_chosen(edt)
         write_global_compat_info(edt)
 
+    with open(args.device_header_out, "w", encoding="utf-8") as dev_header_file:
```
Would you mind making a separate function for this?

galak replied:
Done.

mbolivar-nordic commented:
This seems to make DEVICE_DT_DECLARE() more of an implementation detail than API users need to know about, wouldn't you say @pabigot ?

pabigot commented:
> This seems to make DEVICE_DT_DECLARE() more of an implementation detail than API users need to know about

I suppose so, yes.  In which case, if this gets in, we'll need to update the documentation again.  Or perhaps that could be part of this PR....

mbolivar-nordic commented:
>  In which case, if this gets in, we'll need to update the documentation again.

I'll handle it this time.

mbolivar-nordic commented:
I think that means this PR should rename DEVICE_DT_DECLARE to Z_DEVICE_DT_DECLARE and remove all the users in the tree too.

galak committed:
device: Remove DEVICE_DT_DECLARE / DEVICE_DT_INST_DECLARE

Now that we generate a header that extern's all possible devicetree
based device struct we can remove DEVICE_DT_DECLARE and
DEVICE_DT_INST_DECLARE as they aren't needed anymore.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic commented on include/device.h:
```diff
@@ -767,6 +767,9 @@ static inline int device_pm_put_sync(const struct device *dev) { return -ENOTSUP
 }
 #endif
 
+/* device_extern is generated base on devicetree nodes */
```
"based on" or "from"

galak replied:
fixed.

pabigot replied:
s/base/based/

mbolivar-nordic reviewed (approved):
The device.h comment has a typo. Looks good otherwise.
